nl80211: take hostapd ctl iface path from agent config file

MuteSpirit opened the pull request:
During porting prplMesh to RDKB on RPi it has been founded that
prplMesh assumptions for hostapd ctrl ifaces paths are not
valid for RPi.

Instead of "/var/run/hostapd/<wlan-iface-name>" the paths are
```
    /var/run/hostapd0/wlan0
    /var/run/hostapd1/wlan1
    /var/run/hostapd4/wlan2
```
Add variables for hostapd and wpa supplicant ctrl ifaces for
each agent at beerocks_agent.conf:

    hostap_ctrl_iface
    wpa_supplicant_ctrl_iface

Default values were kept the same:
```
    [agent0]
    ...
    hostap_ctrl_iface=/var/run/hostapd/wlan0
    wpa_supplicant_ctrl_iface=/var/run/wpa_supplicant/wlan0

    [agent1]
    ...
    hostap_ctrl_iface=/var/run/hostapd/wlan2
    wpa_supplicant_ctrl_iface=/var/run/wpa_supplicant/wlan2

    [agent2]
    ...
    hostap_ctrl_iface=/var/run/hostapd/wlan4
    wpa_supplicant_ctrl_iface=/var/run/wpa_supplicant/wlan4
```

https://jira.prplfoundation.org/browse/PPM-232

arnout reviewed (dismissed):
I don't like the approach taken here. I think this configuration should come from the bpl. In addition, I think it should come in a generic way - the concept of a wpa_ctrl socket is specific for hostapd, so a bwl implementation based on something else may have a different concept. So I think the proper approach would be to have a bpl_cfg function to get the bwl configuration for a radio.

That said, it can probably be done as part of the bpl refactorin task that is planned, so this can be merged as is I suppose.

arnout commented on common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp:
```diff
@@ -115,7 +115,7 @@ static uint8_t wpa_bw_to_beerocks_bw(const std::string &chan_width)
                                          hal_conf_t hal_conf)
     : base_wlan_hal(bwl::HALType::AccessPoint, iface_name, IfaceType::Intel, callback, hal_conf),
-      base_wlan_hal_nl80211(bwl::HALType::AccessPoint, iface_name, callback, BUFFER_SIZE)
+      base_wlan_hal_nl80211(bwl::HALType::AccessPoint, iface_name, callback, BUFFER_SIZE, hal_conf)
```
The same change has to be made in mon_wlan_hal_nl80211 and sta_wlan_hal_nl80211. Note that those don't take a hal_conf argument at the moment, so something like that needs to be added.

MuteSpirit replied:
Thanks for your comment
I now know what classes needs WPA supplicant ctrl iface pathes
I'll mark this PR with "don't merge" until fix this issue

VladyslavTupikin replied:
fixed

MuteSpirit commented on agent/src/beerocks/fronthaul_manager/monitor/monitor_thread.cpp:
```diff
@@ -64,9 +64,13 @@ monitor_thread::monitor_thread(const std::string &slave_uds_, const std::string
 
+    bwl::hal_conf_t hal_conf;
+    // FIXME: got ctrl iface path from [agent0] instead of section according "monitor_iface_"
+    hal_conf.wpa_ctrl_path = beerocks_slave_conf.wpa_supplicant_ctrl_iface[0];
```
It's needed to take not wpa supplicant ctrl path but hostapd ctrl path

VladyslavTupikin replied:
fixed

Commit:
prplMesh: CMakeList: Add new variables for ctrl interfaces

Need to add next variables for fixing RDKB builds:
BEEROCKS_HOSTAP_WLAN1_CTRL_IFACE
BEEROCKS_HOSTAP_WLAN2_CTRL_IFACE
BEEROCKS_HOSTAP_WLAN3_CTRL_IFACE
BEEROCKS_WPA_SUPPLICANT_WLAN1_CTRL_IFACE
BEEROCKS_WPA_SUPPLICANT_WLAN2_CTRL_IFACE
BEEROCKS_WPA_SUPPLICANT_WLAN3_CTRL_IFACE

Also add usage of those variables in
beerocks_agent.conf.in file.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
common: bcl: Add variables for ctrl_iface in sConfigSlave

For reading data from new variables added in
CMakeList.txt need to add variables in sConfigSlave:
hostap_ctrl_iface
wpa_supplicant_ctrl_iface

Also need to update the read_slave_config_file()
method for reading data in new variables.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
agent: slave: Add variables for ctrl iface in sSlaveConfig

Need to add variables in son::slave_thread::sSlaveConfig:
hostap_ctrl_iface and wpa_supplicant_ctrl_iface.

Also update fill_son_slave_config() method for adding
values into the new variables.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
agent: fronthaul_manager: Add wpa_ctrl_path variable

For correct AP Manager configuratuion in the
nl80211 need to add wpa_ctrl_path variable in the
hal_conf_t structure, and modify fronthaul_manager
for creating ap_manager with correct data and ap_manager
for filling up wpa_ctrl_path for hal_conf_t structure.

In fronthaul manager add mechanism for finding correct
wpa_ctrl_path (for AP called hostap_iface_ctrl): need to
chose 1 from 3 given by config file.

In ap_manager modify the constructor for filling up
m_hostap_iface_ctrl field and update create_ap_wlan_hal()
method by adding the m_hostap_iface_ctrl as part of hal_conf_t
structure.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
agent: fronthaul_manager: Modify front_haul manager

For using updated mon_wlan_hal_create() method
need to update monitor_thread constructor by
adding 1 additional string argument and fix
usage of mon_wlan_hal_create() method.

Also need to modify monitor creation in fronthaul
main() function.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: dummy: Modify mon_wlan_hal_dummy constructor

Need to update mon_wlan_hal_create() method
by adding argument for hal_conf_t structure,
and update dummy HAL for using new argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: dwpal: Modify mon_wlan_hal_dwpal constructor

Need to update mon_wlan_hal_create() method
by adding argument for hal_conf_t structure,
and update dwpal HAL for using new argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: nl80211: Modify mon_wlan_hal_nl80211 constructor

Need to update mon_wlan_hal_create() method
by adding argument for hal_conf_t structure,
and update nl80211 HAL for using new argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
agent: backhaul_manager: Update WPA_ATTACH handler

Need to update WPA_ATTACH handler - add reading configuration
for wpa_ctrl_iface from file, find correct path according to
the number of agent, create hal_conf_t structure and tranfer
it to the sta_wlan_hal_create() method.

Also need to update arguments in sta_wlan_hal_create()
method - add new argument for hal_conf_t structure.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: dummy: Modify sta_wlan_hal_dummy constructor

Need to update sta_wlan_hal_create() method
by adding argument for hal_conf_t structure,
and update dummy HAL for using new argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: dwpal: Modify sta_wlan_hal_dwpal constructor

Need to update sta_wlan_hal_create() method
by adding argument for hal_conf_t structure,
and update dwpal HAL for using new argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: nl80211: Modify sta_wlan_hal_nl80211 constructor

Need to update sta_wlan_hal_create() method
by adding argument for hal_conf_t structure,
and update nl80211 HAL for using new argument.

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>
Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
bwl: nl80211: Update initialization for m_wpa_ctrl_path

Need to remove useless condition and change
initialization for m_wpa_ctrl_path variable
from hal_conf_t structure.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

MuteSpirit commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1328,9 +1328,46 @@ bool backhaul_manager::backhaul_fsm_wireless(bool &skip_select)
             // Create a HAL instance if doesn't exists
             if (!soc->sta_wlan_hal) {
+
+                // read slave config file
+                std::string slave_config_file_path = "./" + std::string(BEEROCKS_AGENT) +
+                                                     ".conf"; //search first in current directory
+                beerocks::config_file::sConfigSlave beerocks_slave_conf;
+                if (!beerocks::config_file::read_slave_config_file(slave_config_file_path,
+                                                                   beerocks_slave_conf)) {
+                    slave_config_file_path = BEEROCKS_CONF_PATH + std::string(BEEROCKS_AGENT) +
+                                             ".conf"; // if not found, search in beerocks path
+                    if (!beerocks::config_file::read_slave_config_file(slave_config_file_path,
+                                                                       beerocks_slave_conf)) {
+                        std::cout << "config file '" << slave_config_file_path << "' args error."
+                                  << std::endl;
+                        return 1;
+                    }
+                }
```
I don't understand the reason of read config file here.

1) `backhaul_manager` constructor receives as argument the `config_file::sConfigSlave` object which already contain data from beerocks agent config file
https://github.com/prplfoundation/prplMesh/blob/c70a5977e8e57a1a8ce6c9b24152381161b186c3/agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp#L228

2) You may extend `struct SBackhaulConfig` with array of wpa supplicant ctrl paths and copy them from `sConfigSlave` object

VladyslavTupikin replied:
I created a separate [issue](https://jira.prplfoundation.org/browse/PPM-232) for fixing commentaries

MuteSpirit commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1328,9 +1328,46 @@ bool backhaul_manager::backhaul_fsm_wireless(bool &skip_select)
+                }
+
+                bwl::hal_conf_t hal_conf;
+                hal_conf.wpa_ctrl_path = beerocks_slave_conf.hostap_ctrl_iface[iface_num];
```
It's needed to use WPA Supplicant ctrl iface path instead of Hostapd one for creation STA hal

VladyslavTupikin replied:
I created a separate [issue](https://jira.prplfoundation.org/browse/PPM-232) for fixing commentaries

arnout replied:
Not much to fix here... It's the sucky bwl interface definition. It is called "wpa_ctrl_path" and it's used for the supplicant path in sta_wlan_hal and the hostapd path in mon_ and ap_wlan_hal.

MuteSpirit commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1328,9 +1328,46 @@ bool backhaul_manager::backhaul_fsm_wireless(bool &skip_select)
+                    }
+                }
+
+                // Find correct interface number for get correct path for hostap_ctrl
+                beerocks::bpl::BPL_WLAN_IFACE interfaces[beerocks::IRE_MAX_SLAVES] = {0};
+                int num_of_interfaces = beerocks::IRE_MAX_SLAVES;
+                if (beerocks::bpl::cfg_get_all_prplmesh_wifi_interfaces(interfaces,
+                                                                        &num_of_interfaces)) {
+                    std::cout << "failed to read interfaces map" << std::endl;
+                    return 1;
+                }
+
+                uint8_t iface_num = 0;
+                for (iface_num = 0; iface_num < beerocks::IRE_MAX_SLAVES; iface_num++) {
+                    if (interfaces[iface_num].ifname == iface) {
+                        break;
+                    }
+                }
```
Correlation between iface name and slave num is constant among process life.
Maybe construct `std::map<std::string, uint8_t> slave_iface_idx` in `backhaul_manager` class constructor and use it in this place:
```
iface_num = slave_iface_idx[iface]
```
?

VladyslavTupikin replied:
I created a separate [issue](https://jira.prplfoundation.org/browse/PPM-232) for fixing commentaries

arnout replied:
Note that all this also disappears if we have `bpl_cfg_get_wpa_supplicant_ctrl_path`.

MuteSpirit commented on common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.h:
```diff
@@ -12,6 +12,8 @@
 #include "base_wlan_hal_nl80211.h"
 #include <bwl/sta_wlan_hal.h>
 
+#define BUFFER_SIZE 4096
```
To be consistent with nl82011 sta and mon HALs it's needed to define this macro in file `common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp` and inside namespace `bwl::nl80211`

VladyslavTupikin replied:
I created a separate [issue](https://jira.prplfoundation.org/browse/PPM-232) for fixing commentaries

arnout replied:
I don't think this BUFFER_SIZE is even used?

arnout commented on agent/src/beerocks/fronthaul_manager/ap_manager/ap_manager_thread.cpp:
```diff
@@ -112,6 +114,7 @@ bool ap_manager_thread::create_ap_wlan_hal()
 
     bwl::hal_conf_t hal_conf;
     hal_conf.ap_acs_enabled = acs_enabled;
+    hal_conf.wpa_ctrl_path  = m_hostap_iface_ctrl;
```
It would make a whole lot more sense to me to read the path directly from bpl_cfg here. I.e. add a function `bpl_cfg_get_hostap_ctrl_path(radio_name)` that gets the path for that radio.

However, that is a bit of an architectural change, because up to now we mostly used bpl only through the platform manager thread.

arnout commented on agent/src/beerocks/fronthaul_manager/fronthaul_manager_main.cpp:
```diff
@@ -204,8 +205,24 @@ int main(int argc, char *argv[])
     // Create ap_manager
-    son::ap_manager_thread ap_manager(agent_uds, fronthaul_iface, *g_logger_ap_mananger);
+    son::ap_manager_thread ap_manager(agent_uds, fronthaul_iface, *g_logger_ap_mananger,
+                                      beerocks_slave_conf.hostap_ctrl_iface[iface_num]);
```
I made a comment earlier that this path should be retrieved directly from bpl. Actually, this is a better place for it. I.e.,
```suggestion
                                      bpl_get_hostap_ctrl_iface(fronthaul_iface));
```

The UCI implementation can be pretty simple: it just returns "/var/run/hostapd/" + fronthaul_iface, just like the current code. That's OK because UCI means OpenWrt and there the path is always that. And the linux implementation can read it from the platform_db file.

Note that this means that all the previous commits become redundant, except the first one. You still need the first one to populate the default platform_db. Note that we don't have a prplmesh_db.in file, it has to be set from prplmesh_utils.sh.in.

arnout reviewed (approved):
Remaining comments can be handled in follow-up.

VladyslavTupikin commented:
New PR for this feature is #1524
